sdks/java/io/elasticsearch/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java

```diff
@@ -0,0 +1,635 @@
+      private long currentBatchSizeBytes;
+
+      public WriterFn(Write spec) {
+        this.spec = spec;
+      }
+
+      @Setup
+      public void createClient() throws Exception {
+        restClient = spec.getConnectionConfiguration().createClient();
+      }
+
+      @StartBundle
+      public void startBundle(Context context) throws Exception {
+        batch = new ArrayList<>();
+        currentBatchSizeBytes = 0;
+      }
+
+      @ProcessElement
+      public void processElement(ProcessContext context) throws Exception {
+        String document = context.element();
+        batch.add(String.format("{ \"index\" : {} }%n%s%n", document));
+        currentBatchSizeBytes += document.getBytes().length;
+        if (batch.size() >= spec.getBatchSize()
+            || currentBatchSizeBytes >= (spec.getBatchSizeMegaBytes() * 1024L * 1024L)) {
+          finishBundle(context);
+        }
+      }
+
+      @FinishBundle
+      public void finishBundle(Context context) throws Exception {
+        StringBuilder bulkRequest = new StringBuilder();
+        if (batch.size() > 0) {
+          for (String json : batch) {
+            bulkRequest.append(json);
+          }
+          Response response;
+          String endPoint = String.format("/%s/%s/_bulk",
+                                          spec.getConnectionConfiguration().getIndex(),
+                                          spec.getConnectionConfiguration().getType());
+          HttpEntity requestBody = new NStringEntity(bulkRequest.toString(), ContentType
+              .APPLICATION_JSON);
+          response =
+              restClient.performRequest("POST", endPoint, Collections.<String, String>emptyMap(),
+                                        requestBody, new
+                                            BasicHeader("", ""));
+          JsonObject searchResult = parseResponse(response);
+          boolean errors = searchResult.getAsJsonPrimitive("errors").getAsBoolean();
+          if (errors) {
+              StringBuilder errorMessages = new StringBuilder("Error writing to Elasticsearch, "
+                + "some elements could not be inserted:");
+            JsonArray items = searchResult.getAsJsonArray("items");
+            //some items present in bulk might have errors, concatenate error messages
+              for (JsonElement item:items){
+                JsonObject creationObject = item.getAsJsonObject().getAsJsonObject("create");
+                JsonObject error = creationObject.getAsJsonObject("error");
+                  if (error != null){
+                    String docId = creationObject.getAsJsonPrimitive("_id").getAsString();
+                    errorMessages.append(String.format("%n document id %s : ", docId));
+                      String errorMessage = error.getAsJsonObject("caused_by")
+                                            .getAsJsonPrimitive("reason").getAsString();
+                      errorMessages.append(errorMessage);
```

Comment:
No, actually the delimiter is \n. The output of testWriteWithErrors() for 2 malformed documents is 

Error writing to Elasticsearch, some elements could not be inserted:
 document id AVjASLucVlDMC7b3bq3B : Unexpected character (';' (code 59)): was expecting a colon to separate field name and value
 at [Source: org.elasticsearch.common.io.stream.InputStreamStreamInput@714b6929; line: 1, column: 14]
 document id AVjASLucVlDMC7b3bq3E : Unexpected character (';' (code 59)): was expecting a colon to separate field name and value
 at [Source: org.elasticsearch.common.io.stream.InputStreamStreamInput@3bce5cae; line: 1, column: 14]

The only thing I can get in addition to the error message is the doc id. I cannot get the doc _source because it is not in ES error response. If the user does not provide a doc id (if ES generates one at insertion time) that might not help the user identify failing doc. But this is the best information we can give.